Fix: imread large jp2 images on Windows

Ashod opened the pull request:
Jasper library uses temp files to decode large images. The opening of temp files on Windows was broken.

On Windows, the tmpnam function returns a temp filename in the current directory, which has a prepended backslash '\\'.
This subsequently fails the open function, which fails the decoding of the image.

This patch creates a proper temp filename in the temp folder and makes unlink work by opening the file as short-lived.

Test image committed in test_data separately.

Ashod commented:
The assertion that triggers during the unit-test run is removed in my patch.

Can someone please see what's going on?

mshabunin commented:
On linux slave preinstalled libjasper (1.900.1-14ubuntu3) is used:
```
--     JPEG 2000:                   /usr/lib/x86_64-linux-gnu/libjasper.so (ver 1.900.1)
```
And it crashes during the test for some reason. I'll try to investigate this issue.

Ashod commented:
@mshabunin Interesting. Thanks for the help.

mshabunin commented:
@Ashod, I'm sorry, we can not accept changes in 3rdparty code because it can be overwritten after next library update. It is better to contribute it to jasper library upstream.

Concerning the second part of the PR (unit test), I think it is useful, but we need to fix the crash. 
This problem is already known: http://code.opencv.org/issues/3732
Since it is reproduced only when reading in `CV_LOAD_IMAGE_GRAYSCALE` mode, the best way to fix it is to manually convert image to grayscale in `Jpeg2KDecoder::readData` method instead of using jasper library for this.

Ashod commented:
@mshabunin Thanks. I'll do the conversion manually.

WRT my fix in jasper, I can try contributing there, but they might not accept it. I suspect their code works fine under cygwin and msys, but we use VC, which has the issue (I suspect they don't care much about VC). What are we to do if they don't accept my patch?

Meanwhile, we still have a broken jp2 (for large images,) so I'm not sure why we wouldn't want to patch it ourselves especially now that we have unit-tests that will fail if a newer version breaks it?

mshabunin commented:
Maybe @vpisarev can clarify this question better.

vpisarev commented:
@Ashod, @mshabunin, sorry for delay; there are a few problems with this patch:
1) it does not work on Linux or other systems where we use external libjasper, not the one you patched. Failing builds demonstrate this issue: http://pullrequest.opencv.org/#/summary/. Actually, it means that  loading jp2 as grayscale is buggy not only on Windows, but on Linux too, and your patch is partial.
2) the patched version will be overwritten as soon as we update our copy of libjapser with a newer version. We do it from time to time.
3) even when the jasper team accepts your patch and new patched version comes out, it will take time until it becomes mainstream and we can rely on it.

As Maksim noted, when you load image as color, there is no problem; so I agree with Maksim and would suggest you to implement a workaround rather than patch libjasper.

Ashod commented:
@vpisarev I think there is confusion here.
There are two completely unrelated issues.

I have fixed a problem with how libjasper opens temp files when decoding large jp2 images (on Windows). This has _nothing_ with greyscale at all.

The greyscale issue is yet another issue that I have uncovered thanks to the unit-test that I added. For that Maksim suggested that I open as color and do the conversion myself. Which I will probably do in a separate patch (it's only related to this patch because my greyscale unit-test fails on Linux, which I can revert temporarily).
My patch isn't partial, because it doesn't even try to address the color/grey issue at all, my unit-tests reveal the issue on Linux.

So, leaving the color/grey conversion aside, and returning to my fix for opening large images.

First and foremost, the code I'm modifying is _not_ standard libjasper code. It was modified/committed by Andrey Kamaev on 8/29/2012 3:15:07 PM, Commit hash: d8015fd6ed7664bd178ad3a3c9fd3644c7cfb245. And that's only in this file I'm patching. Looking at the history of 3rdParty/libjasper shows many code changes that will be overwritten when updated from the standard library.

So your concern about overwriting changes is less of an issue because the current libjasper code is already customized and not the official one.

Second, when/if the code is overwritten, my unit-tests will simply fail and point that the update is invalid (but I suspect simply copying libjasper files without the patches in OpenCV will not work anyway). It'd be trivial to see the commits in libjasper directory to figure out that my patch is missing. Even if unit-tests aren't run, nothing will be worse than the current situation, which is that any image larger than 16MB will fail to load on Windows (because of the bug I found/patched).

I hope I answered your points adequately, and hope you will share my concern that OpenCV currently is unable to decode large jp2 files. Thanks.

vpisarev commented:
ah, ok. sorry. I was not aware that libjasper is already patched and your patch is actually for the patch :)
Nevertheless, we can not accept a patch with failing tests. Buildbot should be "all green", now it's not.

Ashod commented:
@vpisarev No worries. I've fixed the color-to-grey conversion issue on Linux, and build is green.

Please note that there is a PR #245 on opencv_extra for the jp2 test images required by the unit-tests.

vpisarev commented:
thanks! :+1:
